testing/async/exclusive_scan/large_indices.cu

```diff
@@ -0,0 +1,239 @@
+#include <thrust/detail/config.h>
+
+#if THRUST_CPP_DIALECT >= 2014
+
+#include <async/test_policy_overloads.h>
+
+#include <async/exclusive_scan/mixin.h>
+
+#include <thrust/device_free.h>
+#include <thrust/device_malloc.h>
+#include <thrust/device_ptr.h>
+#include <thrust/optional.h>
+
+#include <thrust/iterator/detail/device_system_tag.h>
+#include <thrust/iterator/detail/iterator_facade_category.h>
+
+#include <cstdint>
+
+// This test is an adaptation of TestInclusiveScanWithBigIndices from scan.cu.
+
+namespace
+{
+
+// Fake iterator that asserts
+// (a) it is written with a sequence and
+// (b) a defined maximum value is written at some point
+struct assert_sequence_iterator
+{
+  using value_type      = std::int64_t;
+  using difference_type = std::int64_t;
+
+  // Defined for thrust::iterator_traits:
+  using pointer           = value_type*;
+  using reference         = assert_sequence_iterator; // weird but convenient
+  using iterator_category =
+    typename thrust::detail::iterator_facade_category<
+      thrust::device_system_tag,
+      thrust::random_access_traversal_tag,
+      value_type,
+      reference>::type;
+
+  std::int64_t expected{0};
+  std::int64_t max{0};
+  mutable thrust::device_ptr<bool> found_max{nullptr};
+  mutable thrust::device_ptr<bool> unexpected_value{nullptr};
+
+  // Should be called on the first iterator generated.
+  void initialize_shared_state()
+  {
+    found_max        = thrust::device_malloc<bool>(1);
+    unexpected_value = thrust::device_malloc<bool>(1);
+    *found_max        = false;
+    *unexpected_value = false;
+  }
+
+  // Should be called only once on the initialized iterator.
+  void free_shared_state() const
```

Comment:
Why is not a destructor? Leave a note explaining.

Reply:
@griwes

Reply:
Added a note.